aot-maven-plugin/src/main/java/com/dylibso/chicory/experimental/maven/aot/AotGenMojo.java

```diff
@@ -123,12 +129,126 @@ public void execute() throws MojoExecutionException {
+        cu.addImport(UncheckedIOException.class);
+        cu.addImport(Parser.class);
+        cu.addImport(WasmModule.class);
+
+        var clazz = cu.addClass(moduleName, Keyword.PUBLIC, Keyword.FINAL);
+
+        var constr = clazz.addConstructor(Keyword.PRIVATE);
+        constr.createBody();
+
+        var method = clazz.addMethod("load", Keyword.PUBLIC, Keyword.STATIC);
+        method.setType(WasmModule.class);
+        var methodBody = method.createBody();
+
+        var getResource =
+                new MethodCallExpr(
+                        new ClassExpr(parseType(moduleName)),
+                        "getResourceAsStream",
+                        new NodeList<>(new StringLiteralExpr(wasmName)));
+        var resourceVar =
+                new VariableDeclarationExpr(
+                        new VariableDeclarator(new VarType(), "in", getResource));
+
+        var returnStmt =
+                new ReturnStmt(
+                        new MethodCallExpr()
+                                .setScope(new NameExpr("Parser"))
+                                .setName("parse")
+                                .addArgument(new NameExpr("in")));
+
+        var newException =
+                new ObjectCreationExpr()
+                        .setType(parseClassOrInterfaceType("UncheckedIOException"))
+                        .addArgument(new StringLiteralExpr("Failed to load AOT WASM module"))
+                        .addArgument(new NameExpr("e"));
+        var catchIoException =
+                new CatchClause()
+                        .setParameter(
+                                new com.github.javaparser.ast.body.Parameter(
+                                        parseClassOrInterfaceType("IOException"), "e"))
+                        .setBody(new BlockStmt(new NodeList<>(new ThrowStmt(newException))));
+
+        methodBody.addStatement(
+                new TryStmt()
+                        .setResources(new NodeList<>(resourceVar))
+                        .setTryBlock(new BlockStmt(new NodeList<>(returnStmt)))
+                        .setCatchClauses(new NodeList<>(catchIoException)));
+
+        dest.add(packageName, moduleName + ".java", cu);
+    }
+
+    private static byte[] rewriteWasm(byte[] wasmBytes, WasmModule module) {
+        var writer = new WasmWriter();
+        Parser.parseWithoutDecoding(
+                wasmBytes,
+                section -> {
+                    if (section.sectionId() == SectionId.CODE) {
+                        var out = new ByteArrayOutputStream();
+                        int count = module.codeSection().functionBodyCount();
+                        writeVarUInt32(out, count);
+                        for (int i = 0; i < count; i++) {
+                            writeVarUInt32(out, 3); // function size in bytes
+                            writeVarUInt32(out, 0); // locals count
+                            out.write(OpCode.UNREACHABLE.opcode());
```

Comment:
This looks a bit strange, instead of just removing the function section we are substituting the functions with "unreachable" code.
I understand that the validation will fail but I would rather disable the validation(since it has already been running on the original module) and remove an entire section instead of changing its content.

Reply:
I agree it's strange, but having valid modules seems better. Making invalid modules work requires the following:

* Add a new API to `Parser` to skip validation
* Change two places in `Instance` to use `functionTypes.length` instead of `functions.length`
* Change `WasmModule` to skip calling `validateModule()` when validation is disabled

It also prevents using tools such as `wasm2wat` to view the modules. The `--no-check` flag doesn't help.

So if you really feel strongly about using invalid modules, I can change it, but having valid modules seems like the overall cleaner solution.

Reply:
Skipping structural validation of the module as opposed to contents seems like a dangerous path and makes the APIs more complicated. There's a difference between *this function body has an invalid instruction stack height* and *these two linked data structures are mismatched*.

Reply:
What I feel strongly about is:
we are packing in the Jar a wasm binary with the same name as the original, but that the user won't be able to use.

I can see users wasting a lot of time before figuring out an "hidden" function body substitution happened without them realizing.

With this context I'd prefer a fully invalid wasm module that will clearly indicate that "something happened" and it's not the original module.

That said I'm fully open to hearing more alternatives:

- keep the original wasm file(how much do we save?)
- have a mechanism the user should explicitly enable to mangle the file
- have an external tool that will explicitly perform the operation when needed
- add yours :-)

Reply:
The savings are both size and load time (no need to parse or validate the code). Here's some size comparisons with the amount saved:

```
wast2json 1_523_335 ->   634_206 (2.4x)
wat2wasm  1_482_029 ->   631_428 (2.3x)
python   26_549_480 -> 3_064_154 (8.7x)
```

We can mention the mangling in the documentation. I'm doubtful that people will try to use it, since they had to go through the effort of using the plugin to generate the AOT code, which is the whole point.

What if we name it differently, maybe call it `.meta` instead of `.wasm`? Then people won't be tempted to use it.

Reply:
> I'm doubtful that people will try to use it

I'm not that much :slightly_smiling_face: , apart from the chances, narrowing down this issue without knowing about what is happening behind the scenes is going to be difficult.

> What if we name it differently

Ok, this might be a fair compromise, what about `.aot-data` or similar?